src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectionPool/HttpConnectionStack.cs

```diff
@@ -0,0 +1,58 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Threading;
+
+namespace System.Net.Http
+{
+    /// <summary>
+    /// A <see cref="ConcurrentStack{T}"/> with an extra head pointer to opportunistically avoid allocations on pushes.
+    /// In situations where Push/Pop operations frequently occur in pairs (common under steady load), this avoids most allocations.
```

Comment:
Would we be better off then having one (or more such) dedicated fields then falling back to a locked stack?

Reply:
It's way better than [just a locked stack](https://github.com/dotnet/runtime/pull/99364#discussion_r1535460080), but it is a bit worse than the current approach: about -2 % on the HttpClient benchmark with a `lock`, -1.5% with a `SpinLock` and extra padding to split the lock and the head.

Reply:
> pr: 3569k
> lockedStack: 3123k

That suggests we're still frequently hitting the fallback case. How deep does the stack get?

Reply:
Looks like the rate of going to the fallback is (roughly):
- stress: 30%
- httpclient: 45%
- yarp: 70%

With two values before the fallback instead of 1 it's:
- stress: 14%
- httpclient: 25%
- yarp: 60%

Yarp numbers aren't great here, there is a much bigger gap between returning a connection and renting it back, so there's often more than 1 connection on the pool.

This might not be worth the extra logic